Add Kestrel security considerations article (#37328)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/servers/kestrel/security-considerations.md

- **Keep `ForwardLimit` at `1`** unless you have a chain of multiple proxies. Setting it to `null` (unlimited) means any number of `X-Forwarded-For` entries are processed, which can be exploited.
- **Host filtering can be relaxed** if the reverse proxy validates the Host header, but defense-in-depth suggests keeping it enabled.
- **HTTPS redirection may be unnecessary** if the proxy terminates TLS and communicates with Kestrel over a private network. If using forwarded headers then `UseHttpsRedirection()` will noop as it will see the `X-Forwarded-Proto` header as HTTPS, recommend keeping it in that case.

> [!NOTE]
> **Proxy-specific guidance:** The recommendations above are general guidelines. Each reverse proxy has its own configuration model and security features. Consult your proxy's documentation for specifics—for example, see the [YARP documentation](xref:fundamentals/servers/yarp/overview) if you are using YARP as your reverse proxy.

> [!WARNING]

> > The Forwarded Headers middleware can be silently enabled with **no proxy restrictions** by setting the environment variable `ASPNETCORE_FORWARDEDHEADERS_ENABLED=true` (or `DOTNET_FORWARDEDHEADERS_ENABLED=true`, or the configuration key `ForwardedHeaders_Enabled`). When enabled this way, the framework **clears `KnownProxies` and `KnownIPNetworks`**, meaning every upstream IP is trusted. This is far more permissive than what is achievable through the C# API, where `KnownProxies` defaults to loopback only. **Verify this variable isn't set in edge deployments**—check your Dockerfiles, orchestrator configs, CI/CD pipelines, and hosting platform defaults.

Reviewer: ```suggestion
> The Forwarded Headers middleware can be silently enabled with **no proxy restrictions** by setting the environment variable `ASPNETCORE_FORWARDEDHEADERS_ENABLED=true` (or `DOTNET_FORWARDEDHEADERS_ENABLED=true`, or the configuration key `ForwardedHeaders__Enabled`). When enabled this way, the framework **clears `KnownProxies` and `KnownIPNetworks`**, meaning every upstream IP is trusted. This is far more permissive than what is achievable through the C# API, where `KnownProxies` defaults to loopback only. **Verify this variable isn't set in edge deployments**—check your Dockerfiles, orchestrator configs, CI/CD pipelines, and hosting platform defaults.
```

Author: I think it's correct currently.
https://source.dot.net/#Microsoft.AspNetCore/ForwardedHeadersOptionsSetup.cs,22

|---|---|
| `Http1` | HTTP/1.1 only |
| `Http2` | HTTP/2 only |
| `Http3` | HTTP/3 only |
| `Http1AndHttp2` (default with HTTPS) | HTTP/1.1 and HTTP/2 |
| `Http1AndHttp2AndHttp3` | All three protocols |

## Security checklist

Quick reference for production deployments. Each row links to the detailed section.

| Feature | Default | [Edge Server](#edge-server-direct-internet-exposure) | [Behind Proxy](#behind-a-reverse-proxy) | Details |
|---|---|---|---|---|
| [**Host Filtering**](#host-filtering) | All hosts allowed | Configure `AllowedHosts` | Optional (defense-in-depth) | [Security Middleware](#security-middleware) |
| [**HTTPS Redirection**](#https-redirection) | Not enabled | Enable | Depends on proxy config | [Security Middleware](#security-middleware) |
| [**HSTS**](#http-strict-transport-security-hsts) | Not enabled | Enable | Enable | [Security Middleware](#security-middleware) |
| [**Forwarded Headers**](#forwarded-headers) | Not enabled | **Don't enable** | Enable with `KnownProxies` | [Security Middleware](#security-middleware) |
| [**Rate Limiting**](#rate-limiting) | Disabled | Enable | Enable (app-level) | [Security Middleware](#security-middleware) |

> | [**CORS**](#cors) | Disabled | Configure per API (if needed) | Configure per API (if needed) | [Security Middleware](#security-middleware) |

Reviewer 2: do we want to add a section about Antiforgery?

### Advanced TLS configuration

- **`OnAuthenticate`**: A callback providing direct access to `SslServerAuthenticationOptions` per connection for advanced scenarios (cipher suite selection, ALPN negotiation, etc.).
- **`UseTlsClientHelloListener`** (.NET 11+): A connection middleware (on `ListenOptions`) that inspects the raw TLS Client Hello bytes before the handshake begins. This enables custom logic based on the client's TLS capabilities (e.g., SNI-based routing, fingerprinting). It must be called **before** `UseHttps()` in the middleware pipeline. It has its own timeout (default 8 seconds), which is **additive** with the TLS handshake timeout—consider reducing each timeout to keep the total bounded (e.g., 5s + 5s instead of 8s + 10s).
- **`TlsClientHelloBytesCallback`** (.NET 10): A property on `HttpsConnectionAdapterOptions` and `TlsHandshakeCallbackOptions` that provides raw Client Hello inspection. In .NET 11 this property is obsolete in favor of `UseTlsClientHelloListener`.

## HTTP/1.1 request parsing and validation

> Kestrel implements HTTP/1.1 parsing per RFC 9112 with several security-motivated constraints.

Reviewer 2: ```suggestion
Kestrel implements HTTP/1.1 parsing per [RFC 9112](https://www.rfc-editor.org/rfc/rfc9112) with several security-motivated constraints.
```
add link to RFC?

### Request line validation

- **Method**: Only valid HTTP token characters are accepted. Invalid token characters in the method cause rejection.

> - **Request target**: Empty paths and paths starting with `?` or `%` are rejected. Null bytes (`0x00`) are rejected because they can cause issues with native servers when forwarded; non-ASCII bytes (>= `0x80`) are rejected during ASCII string conversion. Kestrel's parser scans for delimiters to find target boundaries but doesn't perform a per-byte VCHAR check (RFC 9112 §4), so bytes in `0x01–0x1F` (excluding `\0`, `\n`, `\r`, space) and `0x7F` (DEL) reach `Request.Path` unchanged. The practical risk is log injection—for example, `0x1B` (ESC) in the path can inject ANSI escape sequences into log viewers. Applications can defend by deploying behind a stricter reverse proxy, using a logger provider that escapes control characters, sanitizing logs before displaying them by escaping control characters.

Reviewer 3: ```suggestion
- **Request target**: Empty paths and paths starting with `?` or `%` are rejected. Null bytes (`0x00`) are rejected because they can cause issues with native servers when forwarded; non-ASCII bytes (>= `0x80`) are rejected during ASCII string conversion. Kestrel's parser scans for delimiters to find target boundaries but doesn't perform a per-byte VCHAR check (RFC 9112 §4), so bytes in `0x01–0x1F` (excluding `\0`, `\n`, `\r`, space) and `0x7F` (DEL) reach `Request.Path` unchanged. The practical risk is log injection—for example, `0x1B` (ESC) in the path can inject ANSI escape sequences into log viewers. Applications can defend against this with one or more of the following measures: deploying behind a stricter reverse proxy, using a logger provider that escapes control characters, or sanitizing logs before they're displayed.
```
Is this meant to be one or more?

2. **The final Transfer-Encoding must be `chunked`.** If Transfer-Encoding is present but the final coding isn't `chunked`, the request is rejected. This prevents ambiguous body-framing.
3. **HTTP/2 forbids chunked Transfer-Encoding** per RFC 9113 Section 8.1. Kestrel rejects HTTP/2 requests with chunked encoding.

```
# Rejected: Transfer-Encoding is present but final coding is not chunked
Transfer-Encoding: gzip
-> 400 Bad Request

# Accepted: Content-Length removed, body read as chunked
Transfer-Encoding: chunked
Content-Length: 42
-> Content-Length removed, X-Content-Length: 42 added

# Accepted by Kestrel's HTTP/1 parser: upgrade requests can still carry a body
# and are not rejected solely because body framing headers are present.
Connection: upgrade
Transfer-Encoding: chunked
-> Request accepted; body handling is application/protocol-specific
```

### Chunked encoding validation

Kestrel validates chunked transfer encoding per RFC 9112 Section 7.1:

- Chunk format: `chunk-size [chunk-ext] CRLF chunk-data CRLF`
- Maximum chunk prefix length: 10 bytes (supports chunk sizes up to `0x7FFFFFFF`)

> - Chunk extensions are validated—unpaired `\r` or `\n` in extensions cause rejection

Reviewer 3: ```suggestion
- Chunk extensions are validated—unpaired `\r` or `\n` in extensions cause rejection.
```
Very minor: Complete sentence, so added a period per style guilde.

- `BadChunkSuffix`, `BadChunkSizeData`, `BadChunkExtension`, and `ChunkedRequestIncomplete` errors each produce `400 Bad Request`

An [`EnableInsecureChunkedRequestParsing`](#appcontext-compatibility-switches) AppContext switch exists for backwards compatibility but is disabled by default. See [AppContext Compatibility Switches](#appcontext-compatibility-switches).

> ### Line termination

Reviewer 2: should this section be removed, if it was already described at `Header validation`'s `LF` paragraph? same context switch is mentioned there

HTTP/1.1 requires CRLF (`\r\n`) line terminators. By default, Kestrel also accepts bare LF (`\n`), as permitted by RFC 9112 Section 2.2 ("a recipient MAY recognize a single LF as a line terminator"). This is a robustness accommodation for non-compliant clients and intermediaries. The [`Microsoft.AspNetCore.Server.Kestrel.DisableHttp1LineFeedTerminators`](#appcontext-compatibility-switches) AppContext switch can be set to reject bare LF for defense-in-depth against intermediaries that disagree about framing boundaries.

> 

Author: ```suggestion
```

### Access control

| Option | Default | Description |
|---|---|---|
| `CurrentUserOnly` | `true` | Restricts pipe access to the same user account running Kestrel. |
| Custom `PipeSecurity` |—| Allows fine-grained ACL configuration via `PipeAccessRule` for scenarios requiring access from different users or services. |

### Limitations

- Named pipes are connection-scoped (single stream per connection). HTTP/1.1 and HTTP/2 are supported; **HTTP/3 isn't supported**.
- The concept of graceful vs. ungraceful close doesn't map cleanly to named pipes—Kestrel relies on HTTP-level protocol boundaries (Content-Length, chunked encoding, frames) for message delimitation.

## Unix domain sockets transport security

Kestrel supports Unix domain sockets (UDS) as a transport, commonly used for communication between a reverse proxy (e.g., nginx) and Kestrel on the same machine.

### File permissions

Unix domain sockets are represented as files in the filesystem. Access control is governed by file permissions (owner, group, other). **Kestrel doesn't explicitly set file permissions on the socket file**—the socket inherits permissions based on the process's `umask` at the time of creation.

> For production deployments, ensure the socket file's permissions restrict access to the intended user or group:

Reviewer: ```suggestion
For production deployments, ensure the socket file's permissions restrict access to the intended user or group (`www-data` in the example below):
```
